File: 2024-4-10/Substring with Concatenation of All Words.py
from collections import defaultdict

class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        result = []
        length = len(words[0])
        word_counts = defaultdict(int)
        for word in words:
            word_counts[word] += 1
        for offset in range(length):
            size = 0
            seen = defaultdict(int)
            for i in range(offset, len(s) - length + 1, length):
                sub = s[i:i+length]

                if sub not in word_counts:
                    seen.clear()
                    size = 0
                    continue

                seen[sub] += 1
                size += 1
                while seen[sub] > word_counts[sub]:
                    first = s[i - (size - 1) * length: i - (size - 2) * length]
                    seen[first] -= 1
                    size -= 1
                if size == len(words):
                    result.append(i - (size - 1) * length)

        return result

File: 2024-4-10/Substring with Concatenation of All Words.js
class Solution {
    findSubstring(s, words) {
        const result = [];
        const map = new Map();
        const length = words[0].length;

        map.clear();
        for (const word of words) {
            map.set(word, (map.get(word) || 0) + 1);
        }

        for (let offset = 0; offset < length; ++offset) {
            let size = 0;
            const seen = new Map();
            for (let i = offset; i + length <= s.length; i += length) {
                const sub = s.substr(i, length);

                if (!map.has(sub)) {
                    seen.clear();
                    size = 0;
                    continue;
                }

                seen.set(sub, (seen.get(sub) || 0) + 1);
                size++;
                while (seen.get(sub) > map.get(sub)) {
                    const first = s.substr(i - (size - 1) * length, length);
                    seen.set(first, seen.get(first) - 1);
                    size--;
                }

                if (size === words.length) {
                    result.push(i - (size - 1) * length);
                }
            }
        }

        return result;
    }
}
